service/history/hsm/tree.go

```diff
@@ -185,22 +212,52 @@ func (n *Node) Path() []Key {
 
 // Outputs returns all outputs produced by transitions on this tree.
 func (n *Node) Outputs() []PathAndOutputs {
-	var paos []PathAndOutputs
-	if len(n.cache.outputs) > 0 {
-		paos = append(paos, PathAndOutputs{Path: n.Path(), Outputs: n.cache.outputs})
+	root := n.root()
+	if root.opLog == nil {
+		return nil
 	}
-	for _, child := range n.cache.children {
-		paos = append(paos, child.Outputs()...)
+
+	currentPath := n.Path()
+
+	if root.opLog.IsDeleted(currentPath) {
+		return nil
 	}
+
+	// Group transitions by their path
```

Comment:
Just structure the op log to maintain the information in a way that works as output of this function.

Reply:
I think we are close now. We could do something like, if you prefer:

```
type OperationLog struct {
    // Map of Type -> ID -> []TransitionOutputWithCount
    // This mirrors our node structure and avoids path serialization
    TransitionsByNode map[string]map[string][]TransitionOutputWithCount
    DeletedPaths [][]Key
}
```

Not much perf gain in practice, though